test/tools/unit_test_runner.d

```diff
@@ -0,0 +1,326 @@
+        "-J" ~ projectRootDir.buildPath("res"),
+        "-I" ~ projectRootDir.buildPath("src"),
+        "-I" ~ unitTestDir,
+        "-i",
+        "-g",
+        "-main"
+    ];
+
+    const commonFlags = staticFlags ~
+        testFiles ~
+        runnerPath ~
+        ("-of" ~ outputPath) ~
+        ("-m" ~ model);
+
+    const flags = needsStrtold ? commonFlags ~ ("-L" ~ strtoldObjPath) : commonFlags;
+
+    write(path, flags.join("\n"));
+}
+
+/**
+Returns `true` if any of the given files don't exist.
+
+Also prints an error message.
+*/
+bool missingTestFiles(const string[] givenFiles)
+{
+    const nonExistingTestFiles = givenFiles
+        .map!(file => testPath(file))
+        .filter!(file => !file.exists)
+        .join("\n");
+
+    if (!nonExistingTestFiles.empty)
+    {
+        stderr.writefln("The following test files don't exist:\n\n%s",
+            nonExistingTestFiles);
+
+        return true;
+    }
+
+    return false;
+}
+
+void execute(const string[] args ...)
+{
+    enforce(spawnProcess(args).wait() == 0,
+        "Failed to execute command: " ~ args.join(" "));
+}
+
+void runConfig()
+{
+    enum configPath = projectRootDir.buildPath("config.d");
+    enum outputPath = generatedDir.buildPath("dub");
+    enum versionPath = projectRootDir.buildPath("VERSION");
+
+    auto cmd = [
+        dmdPath,
+        "-m" ~ model,
+        "-run",
+        configPath,
+        outputPath,
+        versionPath
+    ];
+
+    version (Posix)
+        cmd ~= "/etc";
+
+    execute(cmd);
+}
+
+void buildStrtold()
+{
+    if (!needsStrtold)
+        return;
+
+    const cmd = [
+        environment.get("CC", "cl"),
+        "/nologo",
+        "/EHsc",
+        "/TP",
+        "/c",
+        projectRootDir.buildPath("src", "dmd", "backend", "strtold.c"),
+        "/Fo" ~ strtoldObjPath,
+        "/I",
+        projectRootDir.buildPath("src", "dmd", "root")
+    ].join(" ");
+
+    enforce(spawnShell(cmd).wait() == 0, "Failed to execute command: " ~ cmd);
+}
+
+bool needsStrtold()
+{
+    version (Windows)
+    {
+        version (DigitalMars)
+            return model == "32mscoff" || model == "64";
+
+        return true;
+    }
+
+    return false;
+}
+
+int main(string[] args)
+{
+    string unitTestFilter;
+    getopt(args, "filter|f", &unitTestFilter);
+
+    const givenFiles = args[1 .. $];
+
+    if (missingTestFiles(givenFiles))
+        return 1;
+
+    enum runnerPath = resultsDir.buildPath("runner.d");
+    const testFiles = givenFiles.testFiles;
+
+    mkdirRecurse(resultsDir);
+    testFiles
+        .moduleNames
+        .writeRunnerFile(runnerPath, unitTestFilter);
+
+    enum cmdfilePath = resultsDir.buildPath("cmdfile");
+    enum outputPath = resultsDir.buildPath("runner").setExtension(exeExtension);
+    writeCmdfile(cmdfilePath, runnerPath, outputPath, testFiles);
+
+    runConfig();
+    buildStrtold();
+    execute(dmdPath, "@" ~ cmdfilePath);
+    execute(outputPath);
```

Comment:
So I presume that you save the commands to a file because you want to be able to invoke DMD outside of this tool, but why not add `-run` to the command flags above?

Reply:
> So I presume that you save the commands to a file because you want to be able to invoke DMD outside of this tool

The assumption is that there might be a lot of files added and I think Windows has a limitation on how many arguments can be passed to an application.

> but why not add `-run` to the command flags above?

It didn't work. Seems to be some kind of bug. I didn't mange to create a small test case.